eng/scripts/linting_suppression_generator.py

```diff
@@ -0,0 +1,256 @@
+    file_elements: List[ET.Element] = tree.getroot().findall('file')
+
+    for file_element in file_elements:
+        file_name = file_element.attrib['classname']
+
+        errors: List[ET.Element] = file_element.findall('BugInstance')
+        for error in errors:
+            type = error.attrib['type']
+            message = error.attrib['message'].replace('"', '\'').replace('<', '&lt;').replace('>', '&gt;')
+            line_number = error.attrib['lineNumber']
+
+            if type not in violations:
+                violations[type] = set()
+            
+            violations[type].add((file_name, message, line_number))
+
+    if len(violations) == 0:
+        # Don't update the suppression file if there are no violations.
+        # This could indicate that the build failed or was ran incorrectly.
+        print('No Spotbugs violations were found.')
+        return
+    
+    # Now that we have the violations, we can generate the suppression file.
+    # The format of the suppression file is as follows:
+    #
+    # <?xml version="1.0" encoding="UTF-8"?>
+    # <FindBugsFilter xmlns="https://github.com/spotbugs/filter/3.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    #     xsi:schemaLocation="https://github.com/spotbugs/filter/3.0.0 https://raw.githubusercontent.com/spotbugs/spotbugs/3.1.0/spotbugs/etc/findbugsfilter.xsd">
+    #   <Match>
+    #     <Bug pattern="..." />
+    #     <Class name="..." />
+    #   </Match>
+    # </FindBugsFilter>
+    #
+    # If the same bug applies to multiple classes, then the <Class> element can be repeated:
+    #
+    # <Match>
+    #   <Bug pattern="..." />
+    #   <Or>
+    #       <Class name="..." />
+    #       <Class name="..." />
+    #       ...
+    #   </Or>
+    # </Match>
+    #
+    # Before each class an XML comment will be added with the message and line number of the violation.
+    # This is to make it easier to understand what the violation is for, and how to remedy it.
+    with open(file=os.path.join(project_folder, 'spotbugs-exclude.xml'), mode='w') as spotbugs_suppressions:
+        spotbugs_suppressions.write('<?xml version="1.0" encoding="UTF-8"?>\n\n')
+        spotbugs_suppressions.write('<FindBugsFilter xmlns="https://github.com/spotbugs/filter/3.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"\n')
+        spotbugs_suppressions.write('                xsi:schemaLocation="https://github.com/spotbugs/filter/3.0.0 https://raw.githubusercontent.com/spotbugs/spotbugs/3.1.0/spotbugs/etc/findbugsfilter.xsd">\n')
+    
+        for type in violations.keys():
+            spotbugs_suppressions.write('  <Match>\n')
+            bugs = sorted(violations[type], key=lambda x: int(x[2]))
+
+            # If there is only one bug, then we can just write it out.
+            if len(bugs) == 1:
+                bug = bugs.pop()
+                spotbugs_suppressions.write(f'    <!-- Line: {bug[2]}, Message: {bug[1]} -->\n')
```

Comment:
Line number will go stale very quickly, won't it? Might be better to omit it than have it here and confuse future us.

Reply:
This has been removed